### ISSUE TITLE
Specification to capture unmodified audio

### ISSUE THREAD

**Issue by klausj** [w3c/mediacapture-main#457]
Current versions of Firefox (53.0.3) and Chrome (58.0.3029.110) preprocess audio capture streams for WebRTC even if the echoCancellation constraint property is set to false. I do not know what the preprocessing exactly does, but it adds some artifacts to the stream. And if we capture stereo the input streams are mixed together to mono,  preprocessed and the processed stream is duplicated. The recorded channels are always identical.  I could not find a way to capture unmodified audio using Firefox or Chrome, but this is required for some kind of applications. In our case we record speech for phonetical analysis. With this audio artifacts the recordings are not appropriate for our needs. The preprocessing is also problematic for music recording apps.
Surprisingly Microsoft Edge is the only browser I've tested, which does no audio preprocessing.

I think that Media Streams API should specify that implementations should not preprocess audio if echoCancellation is set to false.

**Comment by fluffy** [comment]
I agree there is a need for unprocessed audio. But I would suggest using a different property for this rather than overloading the echoCancellation

**Comment by klausj** [comment]
Yes, indeed, the echoCancellation property was the only preprocessing property so far, but meanwhile the specification has changed. Now there are the additionally properties autoGainControl and noiseSuppression.
It would be helpful if an application could rely on unmodified audio data if it is able to set all of those properties to false. As far as I understand the current spec, this should be possible if the MediaTrackCapabilities object returned by the user agent contains a false value in each of the preprocessing property sequence values (e.g.  sequence<boolean>   autoGainControl).

But generally I would prefer to get unmodified audio as default. The  requirements for preprocessing might change in the future and additional processing tasks might be added (beam forming with microphon arrays for instance). If those new tasks would be enabled by default it would change the audio quality of existing applications.

I'm currently using this constraints to get unmodified audio from Chrome:

```javascript
{
 audio: {
   "mandatory": {
      "echoCancellation": false,
      "googEchoCancellation": "false",
      "googAutoGainControl": "false",
      "googNoiseSuppression": "false",
      "googHighpassFilter": "false"
     }
    },
  video: false
 }
```

The current spec contains all preprocessing properties of Chrome except the highpass filter.
You can take this as an example: If Chrome enables the high pass filter by default, an application would unintentionally capture filtered (and therefore altered)  audio data, even if it sets all preprocessing properties of the spec to false.

**Comment by burnburn** [comment]
It might be too late in the process to add this new feature, but I believe what you need is a new constrainable property along the lines of "disableInputProcessing", which if satisfied would guarantee that no input processing was done regardless of the settings of the other input processing properties, including vendor-supplied ones.  @klausj, would you be willing to create a PR (or new issue) for that?  It might go into the Icebox and wait until the next version of the spec, but we wouldn't lose track of it.  It would be helpful if you considered what the effect on video would be as well, although we could make it an audio-only property if necessary.

**Comment by fluffy** [comment]
Just as FYI on this ... currently there is no way for the browser to get unprocessed audio on IOS devices. I still think adding this is a good idea because there are many situations where you really want unprocessed audio and hopefully this will change in future IOS releases.

**Comment by klausj** [comment]
I think the builtin microphone signal on most mobile devices is already preprocessed (noise reduction, dynamic compression, etc.). Maybe already by the hardware or by the driver or operating system. And I am not aware of any audio interface API (DirectSound, CoreAudio, ALSA/PulseAudio,...) which has the capability to somehow guarantee unprocessed audio or could be queried about pre-processing. So, I suggest that a property to disable input processing should only be related to input processing in the browser itself.
@burnburn, I do not know any builtin signal processing of video which modifies the video stream, but the property could be preventively meant for video tracks as well.
The video might have artifacts caused by lossy compression encoding. But this is probably a different thing and it might be worth to think about another property, for instance "disableLossyCompression". But I don't know if browsers currently do compression encoding of video input at all.

I will try to create a pull request in the next days.

**Comment by burnburn** [comment]
Thanks for the PR!  If we can agree on the wording details (which may end up being vague and hand-wavy on purpose), the concept of this is nice.  One thing you should consider is what tests can/could be written to verify implementation.

**Comment by alvestrand** [comment]
@klausj we need your IPR sign-off (easiest: if your org is a W3C member, join the group)

**Comment by dontcallmedom** [comment]
@alvestrand fwiw I'm in discussion with @klausj on this very topic at the moment

**Comment by alvestrand** [comment]
@fluffy would we then advise to return "overconstrained" if one asks for this constraint on IOS?

**Comment by kdavis45** [comment]
@klausj , @burnburn ,

With respect to the question of whether hardware prefilters audio with no way to disable the filter(s) AND therefore this processing can not be disabled by webRTC (to get raw audio), my experience has been that raw audio is always obtainable natively and therefore there is no reason it wouldn't also be available to the browser (which at its core is native). Does anyone know that there exists hardware that absolutely provide no raw audio mode (and in what platform)? Every iphone, android phone, chromebook, windows pc and macbook I have at my disposal seems to have raw audio sampling (via native code or by setting webRTC audio: constraints:false). Albeit I did have to manually configure a chromebook to raw audio sampling which is really unacceptable that it could not be done via the browser with webRTC.

However, my bigger concern is that we feel the need to add a constraint at all for raw audio. Shouldn't it simply be raw by default? Isn't a constraint by definition constraining something and therefore raw audio is by definition unconstrained? Someone help me understand this and the need for any constraint definition for raw audio. Even if the hardware or codec is by default filtering the audio, the browsers mediacapture audio: code should, in my opinion, start with raw audio and add filters as requested. Am I wrong in assuming that the browsers are running natively (even if on top of IOS Safari).

I have read that certain codecs only have sample rates in multiples of 24000 hz. So my other concern is that arguably the most common audio sample rate of 44.1KHz is not universally available to webRTC development (made even worse by the fact that sample rate is selectable natively and it is available natively on IOS and Android anyway). I think, should 44.1KHz not be available via the hardware and a developer requests it, rather than deny the request, the webRTC handler in the browser should resample 48KHz down to 44.1 behind the scenes from the developer and this should apply to any analyser nodes attached.

**Comment by kdavis45** [comment]
@fluffy 

I am a bit confused by your comment above:

_**Just as FYI on this ... currently there is no way for the browser to get unprocessed audio on IOS devices. I still think adding this is a good idea because there are many situations where you really want unprocessed audio and hopefully this will change in future IOS releases.**_

If I can write native apps and have access to raw audio natively (and I can AFAIK,..at least my high frequency audio signalling above even 20KHz at 44100 sample rate is working which should be the first thing to get disrupted with almost any audio filter), then why is in not possible for a natively written IOS browser to deliver raw (aka unprocessed) audio and pass that on thru its webRTC block? In fact, I believe it should be possible for the webRTC block in the Chrome browser on IOS to deliver not only raw audio full analyzer nodes with FFT sizes == 32768 in IOS 11 even though Safari 11 does not allow fft sizes above 2048. Any native app (including a browser running on top of Safari 11) should have access to native audio as well as IOS core FFT.

Safari 11 will move at Apples whim to comply with webRTC, but what is restricting Chrome or Firefox or any other browser running natively on an iphone from building a webRTC compliant block including raw audio and full fft analyser nodes. (and dare I say, on a personal note, I believe windowing (i.e. hamming, gaussian) as well as raising the fftsize limit to 65536 should be added to webRTC across the board but that is a different discussion)

**Comment by kdavis45** [comment]
@klausj , @burnburn ,

**_But generally I would prefer to get unmodified audio as default._**

Sorry I did not see this comment from @klausj before now. Yes, I wholeheartedly agree. Even if this is NOT the default state of the HW, it should be the default state of the webRTC software.

**Comment by alvestrand** [comment]
One thought .... the specification of the interface to WebAudio doesn't allow any processing (I think).
Perhaps we should make sure that this stays that way, and drop this?

**Comment by kdavis45** [comment]
@alvestrand, I guess it depends on how you define processing but beyond all the existing filters (below) which actively process the raw audio (by default in most cases), there is also the case of the analyzer node which is attached to the audio stream to perform FFT processing. So I am not sure what you mean when you say the spec does not allow any processing. What kind of processing are you referring to specifically? 

Sample rate should be at the HW processing layer but could be in a firmware core layer if HW can't do it and Windowing of the raw audio is certainly less complicated than most any of the existing filters and certainly FFT. But I would be happy if the default mic stream was simply a raw sample at the two most common sample rates of 44.1KHz and 48 KHz.

echoCancellation: false,
            noiseSuppression: false,
            autoGainControl: false,
            googEchoCancellation: false,
            googAutoGainControl: false,
            googExperimentalAutoGainControl: false,
            googNoiseSuppression: false,
            googExperimentalNoiseSuppression: false,
            googHighpassFilter: false,
            googTypingNoiseDetection: false,
            googBeamforming: false,
            googArrayGeometry: false,
            googAudioMirroring: false,
            googAudioMirroring: false,
            googNoiseReduction: false,
            mozAutoGainControl: false

**Comment by klausj** [comment]
I'm convinced that the best approach would be to have all filters be switched off by default. At least the experimental (prefixed) filters which are not listed in the specification.
Otherwise applications may produce unpredictable audio quality depending on the browser. 
And I think the specification would become clearer.
Besides backwards compatibility, I don't see any drawbacks of this approach.

**Comment by RWOverdijk** [comment]
I apologize in advance if this question is stupid.

Does this issue by any chance have something to do with why when I set up a call through a basic WebRTC implementation, the audio arrives in poor quality? I've rerouted my output to my input as a test (on my Macbook) so I can play music (on spotify for example), and listen to it on a different machine.

It works, but the quality appears mono and low in general, like it's actually being recorded through the microphone (even though I have confirmed locally using photo booth that the quality is recorded properly). I'm trying to achieve the same quality I'd hear on my own machine and I can't seem to find a good resource to make this happen.

**Comment by sourabharsh** [comment]
Guys, I have been trying to read the raw audio data in mobile Chrome for last 3 months. I have tried different different filters but haven't been successful in getting the raw audio data. 
While we work on fixing this in the Chrome , is there a way that I can use to get the raw audio data from the current mobile chrome?

**Comment by jan-ivar** [comment]
I never liked the notion of having a constraint that influenced the default values of other constraints.

https://github.com/w3c/mst-content-hint/pull/31 aims to address https://github.com/w3c/mst-content-hint/issues/28 by making [contentHint](https://w3c.github.io/mst-content-hint/#dom-mediastreamtrack-contenthint) a normative mechanism to set the default values of constrainable audio properties.

In other words:
```js
track.contentHint = "music"
```
would normatively default all speech-focused constrainable audio processing properties to off.

This also offers a framework for defining future audio constraints and their defaults.

I vote we close this issue in favor of https://github.com/w3c/mst-content-hint/issues/28.
